csrc/dynamic_type.h

```diff
@@ -95,6 +98,17 @@
 // DynamicTypeTest.ExamplesInNote, so if you want to change anything in this
 // doc, please make sure to update the test as well.
 //
+// DynamicType also support nested types, that is, the type list can contain
+// DynamicType. For example, we can have:
+//   using IntFloatVecList
+//     = DynamicType<Containers<std::vector, std::list>, int, float>;
```

Comment:
Honestly, I don't like this interface. I prefer to have
```C++
DynamicType<int, float, std::vector, std::list>;
```
Which means `DynamicType` needs to be defined as
```C++
template <typename... Ts, template <typename...> typename... Containers>
struct DynamicType {
```
but his will not compile with compiler complaining that `...Ts` must be the last template argument.

I don't know if this is a limitation of the C++ standard or the compiler, but I don't think this limitation makes sense. Because, `Ts` is a parameter pack of types, `Containers` is a parameter pack of templates. They are clearly different things, and can be distinguished and deducted. This is not like `template<typename... T1s, typename ...T2s>` where `<int, float>` has ambiguous deduction of `T1s = {}, T2s = {int, float}`, `T1s = {int}, T2s = {float}`, or `T1s = {int, float}, T2s = {}`